### PR TITLE
feature: add more vmethods to Lemplate.

### PR DESCRIPTION
**Pull request by ingydotnet**
NOTE: Work in progress.

### CODE REVIEW

File: lib/Lemplate.pm [hunk 1]
```diff
@@ -538,6 +534,39 @@ local function stash_set(stash, k, v, default)
...
 end
 
+_M.vmethods = {
+    join = function(list, delim)
+        if delim == nil then
+            delim = ' '
+        end
+        local out = {}
+        for i = 1, #list, 1 do
+            table.insert(out, list[i])
```
Comment (doujiang24) [inline lib/Lemplate.pm]: better use `out[i * 2 - 1] = list[i]`?

File: lib/Lemplate.pm [hunk 2]
```diff
@@ -538,6 +534,39 @@ local function stash_set(stash, k, v, default)
...
+            if i ~= #list then
+                table.insert(out, delim)
```
Comment (doujiang24) [inline lib/Lemplate.pm]: better use `out[i * 2] = delim`?

**Review by doujiang24 — COMMENTED** [review]
@ingydotnet Thanks, these methods works on my side now :)

File: lib/Lemplate.pm (hunk 1 above)
Comment (agentzh) [reply lib/Lemplate.pm]: Agreed. We should avoid using `table.insert` since it is an `O(n)` operation due to the `lj_tab_len` call inside its implementation.

File: lib/Lemplate.pm [hunk 3]
```diff
@@ -538,6 +534,39 @@ local function stash_set(stash, k, v, default)
...
 
+_M.vmethods = {
+    join = function(list, delim)
+        if delim == nil then
+            delim = ' '
+        end
+        local out = {}
+        for i = 1, #list, 1 do
+            table.insert(out, list[i])
+            if i ~= #list then
```
Comment (agentzh) [inline lib/Lemplate.pm]: We should avoid using `#` on tables in hot loops since it will involve the `lj_tab_len` primitive which is a expensive `O(n)` operation, just like `table.insert`.

Comment (agentzh) [reply lib/Lemplate.pm]: BTW, use of `#` on Lua strings is fine, since it is an `O(1)` operation.

File: lib/Lemplate.pm [hunk 4]
```diff
@@ -538,6 +534,39 @@ local function stash_set(stash, k, v, default)
...
+                table.insert(out, delim)
+            end
+        end
+        return table.concat(out)
```
Comment (agentzh) [inline lib/Lemplate.pm]: We should avoid such table lookups on hot code paths. Better cache the result of the table lookup of the field `concat` in table `table` in the outer scope. Please fix other similar places.

File: lib/Lemplate.pm [hunk 5]
```diff
@@ -538,6 +534,39 @@ local function stash_set(stash, k, v, default)
...
+    end,
+
+    first = function(list)
+        return list[1]
+    end,
+
+    push = function(list, elem)
+        table.insert(list, elem)
```
Comment (agentzh) [inline lib/Lemplate.pm]: Just a side note: we can avoid this `O(n)` operation by caching the table length ourself inside the table, just like how we are handling perl 6 arrays in the fanlang compiler. But we can do such optimizations in `edgelang-fan` instead of here.

**Comment by ingydotnet** [comment]
The lemplate `stash_get` function is missing several edge cases. Fixing that now with tests.

**Comment by ingydotnet** [comment]
This is done afaict. Please review.

**Comment by agentzh** [comment]
@ingydotnet Merged. Thanks!
